common/lib/client/realtimepresence.js

```diff
@@ -402,11 +403,20 @@ var RealtimePresence = (function() {
 		var event = args[0];
 		var listener = args[1];
 		var callback = args[2];
+		var channel = this.channel;
 		var self = this;
 
-		waitAttached(this.channel, callback, function() {
-			self.subscriptions.on(event, listener);
-		});
+		if(channel.state === 'failed') {
+			callback(ErrorInfo.fromValues(RealtimeChannel.invalidStateError('failed')));
+			return;
+		}
+
+		this.subscriptions.on(event, listener);
+		if(callback) {
+			channel.attach(callback);
+		} else {
+			channel.autonomousAttach();
```

Comment:
Why this? (I realise this predated this PR.)

It's no different from attach except that it doesn't call the callback. Why not in `attach()` simply have `callback || noop` or `callback = callback || (err) => err && Logger......` ?